### ISSUE TITLE
[Feature Request] Generic OIDC/OAuth2 support

### ISSUE THREAD

**Issue by hazzuk** [crocodilestick/Calibre-Web-Automated#90]
**Is your feature request related to a problem? Please describe.**
Calibre-web only supports SSO OAuth with external providers Google and Github.

**Describe the solution you'd like**
Adding a generic implementation would allow for authentication with any provider, most importantly allowing self-hosted solutions like [Authentik](https://github.com/goauthentik/authentik), [Authelia](https://github.com/authelia/authelia) and [Keycloak](https://github.com/keycloak/keycloak).

**Additional context**
This has been a topic of discussion for many years on Calibre-web, but the maintainer has said quite definitively that ["We will not extend Calibre-Web with any more login abilities"](https://github.com/janeczku/calibre-web/issues/2965#issuecomment-1875551176).

A [Pull Request](https://github.com/janeczku/calibre-web/pull/2211) was even made with this functionality, but sadly was never merged.

**Comment by crocodilestick** [comment]
This is super interesting but something I'm not particularly versed in. I'm stretched pretty thin atm but if someone can make a PR for CWA using the implementation mentioned in the PR you linked I'll include it in the next release ASAP.

**Comment by Eschguy** [comment]
With #132 being closed, does that mean this feature is incoming?

**Comment by burgerga** [comment]
Closed means not merged, so no :(

**Comment by tecosaur** [comment]
I'm interested in adding CWA to the collection of apps I'm self-hosting, the lack of OIDC support adds a fair bit of hassle when everything else manages to hook into the same auth system. 

I think it's also worth noting that with the rise of self-hosted SSO solutions (Authentik, Zitadel, Keycloak, Authelia, Pocket ID, etc.), while this feature has historically been rather niche, it's now of increasing importance in the self-hosting space.

**Comment by mischavandenburg** [comment]
I would love this to be implemented as well.

I'm running CWA on Kubernetes, so in the meantime, I'm considering to look into an auth sidecar.

**Comment by balgerion** [comment]
OIDC/OAuth2 support an this is absolutely dream version of calibre :)

**Comment by Zyanido** [comment]
Would love to have this integrated!

**Comment by surfingbytes** [comment]
> if someone can make a PR for CWA using the implementation mentioned in the PR you linked I'll include it in the next release ASAP.

The PR is there, closed, but not merged? Is there any info on it why is it not merged? This is feature requested by many, so if we get more info on what's blocking it from merging, someone may update the PR.

**Comment by angriestunicorn** [comment]
Yet another person who would love to have CWA integrated with my Authentik installation. Not a dead subject. Failing to see what the difficulty is in implementing generic oauth.

**Comment by Saetron** [comment]
Getting oauth would be nice, makes login way easier with having a central provider

**Comment by burgerga** [comment]
Yes, we all like OIDC/Oauth2, but please just use the thumbs up on the issue to indicate this, no need to flood everyone's inbox with  "me too"s . 

If you:
* want to know why the linked PR was not merged, have a look at it, I think it's fairly obvious (it didn't work)
* are wondering what the difficulty is in implementing it, please create a PR and show us how easy it is

**Comment by frankvoelker** [comment]
Any ideas why I cant even use the normal LDAP authentication? I follow tutorial here:

https://integrations.goauthentik.io/integrations/services/calibre-web/

and it sais "0 User Successfully Imported". but obviously there are no users imported.

does this LDAP "SSO" mean, that I can just import users via CWA admin UI and then if a user changes his password in the authentication provider it still has the old password in CWA so I would have to re-import users every day or so?

**Comment by Buco7854** [comment]
> Any ideas why I cant even use the normal LDAP authentication? I follow tutorial here:
> 
> https://integrations.goauthentik.io/integrations/services/calibre-web/
> 
> and it sais "0 User Successfully Imported". but obviously there are no users imported.
> 
> does this LDAP "SSO" mean, that I can just import users via CWA admin UI and then if a user changes his password in the authentication provider it still has the old password in CWA so I would have to re-import users every day or so?

LDAP works fine with CWA.
Can't really help without knowing your setup in Authentik, it's probably about groups filtering I would guess, but it's not really the place for that.

Regarding the password it should be managed by authentik. If you edit the password in Calibre Web it will just be ignored, it will not edit it in the LDAP server and it will not work to login. 
Here is my config if it can help.

Login type:
LDAP
LDAP Server Host Name or IP Address:
security-server-1.lan
LDAP Server Port
389
LDAP Encryption:
None
LDAP Authentication:
Simple
LDAP Administrator Username:
cn=calibre_web_service,ou=users,ou=calibre_web,dc=example,dc=lan
LDAP Distinguished Name (DN)
ou=calibre_web,dc=example,dc=lan
LDAP User Object Filter:
(&(objectclass=user)(cn=%s))
LDAP Server is OpenLDAP?:
YES (Should be set with authentik I, I think I used to have some issue cause of that)
LDAP Group Object Filter:
(&(objectclass=group)(cn=%s))
LDAP Group Name:
calibre_web_user
LDAP Group Members Field:
member
LDAP Member User Filter Detection:
Custom Filter
LDAP Member User Filter:
(cn=%s)

In Authentik:
Everytime I have an app that needs LDAP I create an Authentik Application for the LDAP that I link to a Provider for this app. (even if I already have it for like OIDC)
I have a unique ou for it in root DN for the provider.

For Calibre web Provider:

DN:
ou=calibre_web,dc=example,dc=lan

then you need to make sure your user for Calibre Web Service have Search Full LDAP and Can view LDAP permission in the provider.
And that it also have permission to  access the Application for it.

As said before it's not the place to ask about LDAP so please avoid replying using this canal.

**Comment by tseho** [comment]
Hello,
I've opened a new pull request #590 to add this feature.
